### PR TITLE
EI-279 Update VRF Case Status for Legal Entity

### PR DESCRIPTION
**Pull request by ruslanss**

### CODE REVIEW

File: src/SFA.DAS.EmployerIncentives.Database/Scripts/Script.PreDeployment1.sql [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿/*
```
Comment (David-B-Read) [inline src/SFA.DAS.EmployerIncentives.Database/Scripts/Script.PreDeployment1.sql:1]: Is this file needed?

File: src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/Features/UpdateVrfCaseDetailsForNewApplications.feature [hunk 2]
```diff
@@ -1,4 +1,5 @@
-﻿@database
+﻿@ignore
```
Comment (David-B-Read) [inline src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/Features/UpdateVrfCaseDetailsForNewApplications.feature:1]: Is the ignore set here because the tests won't run on CI?

**Review by David-B-Read — APPROVED** [review]

File: src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/DataAccess.cs [hunk 3]
```diff
@@ -23,12 +23,21 @@ public Account GetAccountByLegalEntityId(long legalEntityId)
...
+        public Account GetAccountByAccountLegalEntityId(long accountId, long accountLegalEntityId)
+        {
+            using var dbConnection = new SqlConnection(_connectionString);
+            var account = dbConnection.QuerySingle<Account>("SELECT * FROM Accounts WHERE Id = @accountId AND AccountLegalEntityId = @AccountLegalEntityId",
+                new { accountId, accountLegalEntityId });
+
+            return account;
+        }
+
         public void SetupAccount(Account account)
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/DataAccess.cs:35]: Can you use Dapper contrib as in EI-49?  It's a lot simpler I think.

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/DataAccess.cs (hunk 3 above)
Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/DataAccess.cs:35]: hmm this is a test code , can refactor once EI-49 is merged?

File: src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/Steps/UpdateVrfCaseStatusForLegalEntitySteps.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
+﻿using FluentAssertions;
+using SFA.DAS.EmployerIncentives.Data.Models;
+using System;
+using System.Net;
+using System.Threading.Tasks;
+using TechTalk.SpecFlow;
+
+namespace SFA.DAS.EmployerIncentives.Api.AcceptanceTests.Steps
+{
+    [Binding]
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Api.AcceptanceTests/Steps/UpdateVrfCaseStatusForLegalEntitySteps.cs:10]: Please add a 

[Scope(Feature = "... ")]  attribute

File: src/SFA.DAS.EmployerIncentives.Api/Controllers/AccountCommandController.cs [hunk 5]
```diff
@@ -37,5 +38,12 @@ public async Task<IActionResult> SignAgreement([FromRoute] long accountId, [From
             await SendCommandAsync(new SignLegalEntityAgreementCommand(accountId, accountLegalEntityId, request.AgreementVersion));
             return NoContent();
         }
+
+        [HttpPatch("/legalentities/{hashedLegalEntityId}/vendorregistrationform/status")]
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Api/Controllers/AccountCommandController.cs:42]: We shouldn't have path like this.  The path locates the resource

/legalentities/{hashedLegalEntityId}

the patched data is in the body.  So it's a REST interface rather than a RPC call.

The body should details what has changed e.g. as per https://en.wikipedia.org/wiki/JSON_Patch

So in this example would be 

[
    { "op": "replace", "path": "/vendorregistrationform/status", "value": "newStatus" }
]

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Commands/Persistence/IAccountDomainRepository.cs [hunk 6]
```diff
@@ -1,9 +1,13 @@
 ﻿using SFA.DAS.EmployerIncentives.Abstractions.Domain;
 using SFA.DAS.EmployerIncentives.Domain.Accounts;
+using System.Collections.Generic;
+using System.Threading.Tasks;
 
 namespace SFA.DAS.EmployerIncentives.Commands.Persistence
 {
     public interface IAccountDomainRepository : IDomainRepository<long, Account>
     {
+        Task<IEnumerable<Account>> GetByLegalEntityId(long legalEntityId);
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Commands/Persistence/IAccountDomainRepository.cs:10]: These shouldn't be on the domain respository.  Use the IQueryRepository to get data then look up the domain entity using the IDomainRepository.Find (by entity Id).

The entity Id is the key to the domain model.  It's the logical write model it's not designed for querying.

**Review by Stephek — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Commands/Persistence/AccountDomainRepository.cs [hunk 7]
```diff
@@ -24,12 +26,26 @@ public async Task<Account> Find(long id)
...
             if (aggregate.IsNew)
-            {               
+            {
                 return _accountDataRepository.Add(aggregate.GetModel());
             }
 
             return _accountDataRepository.Update(aggregate.GetModel());
         }
+
+        public async Task<IEnumerable<Account>> GetByLegalEntityId(long legalEntityId)
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Commands/Persistence/AccountDomainRepository.cs:36]: Please remove form the Domain repository. Use the existing query repository to find the account Id by other fields.  Then get the entity from the domain repository,  Treat them as separate read/write models with separate concerns.

File: src/SFA.DAS.EmployerIncentives.Commands/UpsertLegalEntity/UpsertLegalEntityCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+        public UpsertLegalEntityCommandHandler(IAccountDomainRepository domainRepository, IHashingService hashingService)
+        {
+            _domainRepository = domainRepository;
+            _hashingService = hashingService;
+        }
+
+        public async Task Handle(UpsertLegalEntityCommand command, CancellationToken cancellationToken = default)
+        {
+            var account = await _domainRepository.Find(command.AccountId);
```
Comment (Stephek) [inline src/SFA.DAS.EmployerIncentives.Commands/UpsertLegalEntity/UpsertLegalEntityCommandHandler.cs:23]: Would be better if this published the AddCommand or updatedCommand depending on if the account exists.  The work don in EI-49 would make this easier to implement.

File: src/SFA.DAS.EmployerIncentives.Commands/Persistence/IAccountDomainRepository.cs (hunk 6 above)
Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Commands/Persistence/IAccountDomainRepository.cs:10]: For what benefit?  It'll slow down the process and bloat the code.

File: src/SFA.DAS.EmployerIncentives.Commands/Persistence/AccountDomainRepository.cs (hunk 7 above)
Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Commands/Persistence/AccountDomainRepository.cs:36]: For what benefit?  It'll slow down the process and bloat the code.  Not being able to get a domain object by anything other than the primary key is a step too far imo.

File: src/SFA.DAS.EmployerIncentives.Api/Controllers/AccountCommandController.cs (hunk 5 above)
Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Api/Controllers/AccountCommandController.cs:42]: This is an interesting one, you're definitely right about it not being RESTful but what would the implementation of a generic patch method look like?  Is it worth taking this as tech debt with a spike to see how we'd implement the patch method without it becoming a bag of spanners?

Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Api/Controllers/AccountCommandController.cs:42]: Yes a tech debt spike would be useful as the patch approach should be applied elsewhere in a consistent manor.

File: src/SFA.DAS.EmployerIncentives.Commands/Persistence/AccountDomainRepository.cs (hunk 7 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Commands/Persistence/AccountDomainRepository.cs:36]: Because the domain repository is purely for hydrating and dehydrating domain objects form the logical write model.  Querying data for other purposes should be done through the existing query interface so that interface can be used elsewhere and indexed appropriately.  It keeps good separation of concerns and prevents the code from becoming a big mess.  It's a pattern following CQRS and although we don't have physical read/write models taking this approach removes the logical distinction.  However as this is supposedly a throw away project then if you want to add querying to this interface then I will resolve this.

Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Commands/Persistence/AccountDomainRepository.cs:36]: It's not querying per-se tho, it's merely retrieval of a domain object to update isn't it?

File: src/SFA.DAS.EmployerIncentives.Commands/Persistence/IAccountDomainRepository.cs (hunk 6 above)
Comment (Stephek) [reply src/SFA.DAS.EmployerIncentives.Commands/Persistence/IAccountDomainRepository.cs:10]: Because this interface is just for hydrating and persisting a single domain object from the logical write store.

There is already an existing IQueryRepository to retrieve data so that should be used as that is it's concern then publish the commands to do the work on each individual domain object.  That would be more performant since it will could handle multiple domain objects in parallel each observing he existing locking strategy.

By adding a query to this interface encourages batch processing.

File: src/SFA.DAS.EmployerIncentives.Domain/Accounts/LegalEntity.cs [hunk 9]
```diff
@@ -38,5 +55,25 @@ public void SignedAgreement(int signedAgreementVersion, int minimumRequiredVersi
...
         }
+
+        public void SetHashedLegalEntityId(string hashedId)
+        {
+            Model.HashedLegalEntityId = hashedId;
+        }
+
+        internal void UpdateVendorRegistrationCaseStatus(string caseId, string vendorId, string status, DateTime caseStatusLastUpdatedDate)
+        {
+            if (VrfStatusIsCompleted()) return;
```
Comment (Pauljgraham) [inline src/SFA.DAS.EmployerIncentives.Domain/Accounts/LegalEntity.cs:66]: This logic looks wrong to me. I understood that we are supposed to ignore the incoming update if the new status is "Case request complete" . This logic updates the record with the status "Case request complete" and the all subsequent updates are then ignored.

File: src/tests/SFA.DAS.EmployerIncentives.Commands.UnitTests/UpdateVrfCaseStatusForLegalEntity/Handlers/WhenHandlingUpdateVrfCaseDetailsForLegalEntityCommand.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+            accounts[0].AddLegalEntity(1, legalEntityToBeUpdated);
+            accounts[0].AddLegalEntity(2, _fixture.Create<LegalEntity>());
+            accounts[0].AddLegalEntity(3, _fixture.Create<LegalEntity>());
+
+            accounts[1].AddLegalEntity(4, _fixture.Create<LegalEntity>());
+            accounts[1].AddLegalEntity(5, legalEntityToBeUpdated);
+
+            _mockDomainRepository.Setup(x => x.GetByHashedLegalEntityId(caseCompletedCommand.HashedLegalEntityId)).ReturnsAsync(accounts);
+            await _sut.Handle(caseCompletedCommand);
+
```
Comment (Pauljgraham) [inline src/tests/SFA.DAS.EmployerIncentives.Commands.UnitTests/UpdateVrfCaseStatusForLegalEntity/Handlers/WhenHandlingUpdateVrfCaseDetailsForLegalEntityCommand.cs:91]: This is proving that after the status update to "Case request complete" we can no longer update the record. I think this is incorrect.

**Review by Pauljgraham — CHANGES_REQUESTED** [review]

File: src/SFA.DAS.EmployerIncentives.Domain/Accounts/LegalEntity.cs (hunk 9 above)
Comment (MichaelYoung1981) [reply src/SFA.DAS.EmployerIncentives.Domain/Accounts/LegalEntity.cs:66]: The logic is correct, it should set it to complete once and then ignore all future updates.

Comment (ruslanss) [reply src/SFA.DAS.EmployerIncentives.Domain/Accounts/LegalEntity.cs:66]: Yes, however we agreed to remove VendorID from the status update. My bad i shouldn't have added it in the first place.

**Review by Stephek — APPROVED** [review]

**Review by Pauljgraham — APPROVED** [review]
